### PR TITLE
Report missing checksums for Go modules

### PR DESCRIPTION
**Pull request by chmeliik**
STONEBLD-1585

For each main module, compare the list of dependency modules with the
list of modules found in its go.sum file. If any *non-local* dependency
is missing in the go.sum file, report that.

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] Docs updated (if applicable)
- [x] Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

**Comment by chmeliik** [comment]
I will add integration tests later or in a separate PR. PTAL

File: cachi2/core/package_managers/gomod.py [hunk 1]
```diff
@@ -537,29 +596,63 @@ def go_list_deps(pattern: Literal["./...", "all"]) -> Iterator[ParsedPackage]:
...
 
-    return main_module, all_modules, all_packages
+    return ResolveResult(main_module, all_modules, all_packages, modules_in_go_sum)
+
+
+def _parse_go_sum(module_dir: RootedPath) -> frozenset[ModuleID]:
+    """Return the set of modules present in the go.sum file in the specified directory.
+
+    A module is considered present if the checksum for its .zip file is present. The go.mod file
+    checksums are not relevant for our purposes.
```
Comment (chmeliik) [inline cachi2/core/package_managers/gomod.py:606]: > go.mod file checksums are not relevant

Not entirely true, but that's a long story.

Go uses the .mod files to do dependency resolution. If we're not sure that all the .mod files had their checksums verified, we can't be sure that the resolved dependencies are the same as they were last time.

But Go also looks at the .mod files of modules that are not needed, while we aim to report only those that are. So even if we took .mod files into account, we wouldn't be able to report whether dependency resolution was 100% checksum-verified.

Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py:606]: And for what it's worth, the .zip file for a module does include its .mod file, so attempts to change the .mod file in the source of a dependency should be caught (but again, only for the modules that are actually needed for the build)

Comment (brunoapimentel) [reply cachi2/core/package_managers/gomod.py:608]: > Go uses the .mod files to do dependency resolution.

Is this why there are some modules in the go.sum file that only have hashes for the go.mod file (and not for the zip file)? [e.g.](https://github.com/redhat-appstudio/build-service/blob/main/go.sum#L2)

File: cachi2/core/package_managers/gomod.py [hunk 2]
```diff
@@ -537,29 +598,63 @@ def go_list_deps(pattern: Literal["./...", "all"]) -> Iterator[ParsedPackage]:
...
+        if not parts:
+            continue
+        if len(parts) != 3:
+            # - when readGoSum (see link above) encounters a malformed line, it returns an error
+            # - every call to readGoSum ignores that error
+            #   - https://github.com/golang/go/blob/d5c5808534f0ad97333b1fd5fff81998f44986fe/src/cmd/go/internal/modfetch/fetch.go#L495
+            #   - https://github.com/golang/go/blob/d5c5808534f0ad97333b1fd5fff81998f44986fe/src/cmd/go/internal/modfetch/fetch.go#L836
+            # - readGoSum modifies a global variable (indirectly, it's passed in as a param)
+            # - => readGoSum will parse the non-broken part of go.sum and then silently fail
+            # - replicate the bug here, because it means that go only uses the non-broken part
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/gomod.py:629]: Ouch! :harold:

I agree that we should have the same behavior as `go`, but maybe this should be something we also report to the user somehow? I feel like this is the type of error that would go unnoticed for the longest of time (and maybe the warning logs here would be unnoticed as well).

Not really sure how to report it... probably a matter for a separate PR, though.

File: cachi2/core/package_managers/gomod.py [hunk 3]
```diff
@@ -184,28 +184,53 @@ def to_component(self) -> Component:
         return Component(name=self.name, purl=self.purl)
 
 
+ModuleID = tuple[str, str]
+
+
+def _get_module_id(module: ParsedModule) -> ModuleID:
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/gomod.py:204]: Nice :+1:

**Review by brunoapimentel — COMMENTED** [review]
Overall, LGTM. Will do a more detailed review later on.

File: cachi2/core/package_managers/gomod.py (hunk 1 above)
Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py:608]: I think so. But I've even seen the opposite case - a module that only has the .zip checksum (and I don't really know what that means - maybe old modules that don't have go.mod)

File: cachi2/core/package_managers/gomod.py (hunk 2 above)
Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py:629]: The only better way to report it would be to raise an exception IMO :slightly_smiling_face:

**Comment by chmeliik** [comment]
Rebased on main

File: cachi2/core/package_managers/gomod.py (hunk 2 above)
Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py:629]: And now that I'm testing things locally, it looks like missing checksums cause `go build` to fail anyway, so reporting missing checksums is probably a bit redundant anyway :sweat_smile:

**Comment by chmeliik** [comment]
Replaced the long explanation of our "broken go.sum" handling with a link to the github issue

File: tests/integration/test_data/gomod_multiple_modules_missing_checksums/fetch_deps_sha256sums.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  "gomod/pkg/mod/cache/download/rsc.io/sampler/@v/list": "sha256:1c6e945d74b4dc17923055e616b0f50f46d05ba0c4d9d080a481370905d3e614",
+  "gomod/pkg/mod/cache/download/rsc.io/sampler/@v/v1.3.0.info": "sha256:7421e51c069e136de8ede93e929238027fd9256f3aa4614f79609835585f9d8d",
+  "gomod/pkg/mod/cache/download/rsc.io/sampler/@v/v1.3.0.lock": "sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
+  "gomod/pkg/mod/cache/download/rsc.io/sampler/@v/v1.3.0.mod": "sha256:a2ae6e8540a40b85e57703cc618b99c5b1075386d958e8999a0e2d4c02f3a705",
+  "gomod/pkg/mod/cache/download/rsc.io/sampler/@v/v1.3.0.zip": "sha256:da202b0da803ab2661ab98a680bba4f64123a326e540c25582b6cdbb9dc114aa",
+  "gomod/pkg/mod/cache/download/rsc.io/sampler/@v/v1.3.0.ziphash": "sha256:ee5ed9d268a6fb84db6e6f237c214c42f2d9ef7a721f4c28083883234ddadb23",
+  "gomod/pkg/mod/cache/download/sumdb/sum.golang.org/lookup/golang.org/x/text@v0.0.0-20170915032832-14c0d48ead0c": "sha256:c73b55d8ff7744fb5468acc1042bd8e135d46810dd2f4349717e8aa8717cc195",
+  "gomod/pkg/mod/cache/download/sumdb/sum.golang.org/lookup/rsc.io/sampler@v1.3.0": "sha256:aea7de67313bc7e4a7bd7b33ee7c0ecfb8ac2c97fa3995849892b128b8b7f690",
```
Comment (chmeliik) [inline tests/integration/test_data/gomod_multiple_modules_missing_checksums/fetch_deps_sha256sums.json:26]: Curious. We end up downloading some stuff from the GOSUMDB when the checksums are missing in go.sum

**Review by chmeliik — COMMENTED** [review]
Added integration test

Currently trying to figure out why Cachi2 reports more internal packages for the same test when the checksums are not missing

```shell
vim -d tests/integration/test_data/gomod_e2e_multiple_modules/bom.json tests/integration/test_data/gomod_multiple_modules_missing_checksums/bom.json
```

**Comment by chmeliik** [comment]
> Currently trying to figure out why Cachi2 reports more internal packages for the same test when the checksums are not missing

Aha.

```shell
$ cd eggs-module/
$ git checkout missing-checksums

$ go list -deps all  | wc -l
/home/acmiel/go/pkg/mod/rsc.io/quote@v1.5.1/quote.go:8:8: missing go.sum entry for module providing package rsc.io/sampler (imported by rsc.io/quote); to add:
	go get rsc.io/quote@v1.5.1
0

$ go list -deps ./... | wc -l
/home/acmiel/go/pkg/mod/rsc.io/quote@v1.5.1/quote.go:8:8: missing go.sum entry for module providing package rsc.io/sampler (imported by rsc.io/quote); to add:
	go get rsc.io/quote@v1.5.1
15
```

```shell
$ git checkout main
$ go list -deps all | wc -l
47

$ go list -deps ./... | wc -l
47
```

So missing checksums also murder our SBOM generation. Cool.

File: tests/integration/test_data/gomod_multiple_modules_missing_checksums/fetch_deps_sha256sums.json (hunk 4 above)
Comment (chmeliik) [reply tests/integration/test_data/gomod_multiple_modules_missing_checksums/fetch_deps_sha256sums.json:26]: ...and the checksums are not stable? Or different in CI compared to my machine?

```
  E             -  'gomod/pkg/mod/cache/download/sumdb/sum.golang.org/lookup/golang.org/x/text@v0.0.0-20170915032832-14c0d48ead0c': 'sha256:c73b55d8ff7744fb5468acc1042bd8e135d46810dd2f4349717e8aa8717cc195',
  E             ?                                                                                                                           ^^^ ^^^^    ^^^  ^    ^^ ^ ---------   ^^^^ --------------------
  E             +  'gomod/pkg/mod/cache/download/sumdb/sum.golang.org/lookup/golang.org/x/text@v0.0.0-20170915032832-14c0d48ead0c': 'sha256:2eb921fff58e7850038432876b8230eab90513672d588eac2976e0b2681489af',
```

Comment (chmeliik) [reply tests/integration/test_data/gomod_multiple_modules_missing_checksums/fetch_deps_sha256sums.json]: Made the check less strict. LMK what you think

**Comment by chmeliik** [comment]
To summarize the current state:

- reporting of missing checksums work as expected
- missing checksums break SBOM generation
- missing checksums make us download not-quite-reproducible data from gosumdb
- missing checksums almost certainly break the user's build later anyway

So I would probably just document the various problems caused by missing checksums. WDYT?

**Comment by brunoapimentel** [comment]
> So missing checksums also murder our SBOM generation. Cool.

So, does the missing data from the `go list` command correspond to the actual modules that have missing checksums? I'm trying to figure out if there is a scenario in which a dependency would be downloaded by Cachi2, but wouldn't be present in the SBOM (since the `go list` output is incomplete).

I'm assuming this won't happen because we merge the ouput of both `go list` and `go download` for module reporting, right?

**Comment by chmeliik** [comment]
> So, does the missing data from the `go list` command correspond to the actual modules that have missing checksums? 

Kind of. I also forgot that we list deps with the `-e` flag, so the effect is something like this

```shell
$ go list -e -deps -json all | jq 'select(.ImportPath | test("rsc.io/sampler"))'
{
  "ImportPath": "rsc.io/sampler",
  "Match": [
    "all"
  ],
  "Incomplete": true,
  "Stale": true,
  "StaleReason": "build ID mismatch",
  "Error": {
    "ImportStack": [
      "rsc.io/sampler"
    ],
    "Pos": "/home/acmiel/go/pkg/mod/rsc.io/quote@v1.5.1/quote.go:8:8",
    "Err": "missing go.sum entry for module providing package rsc.io/sampler (imported by rsc.io/quote); to add:\n\tgo get rsc.io/quote@v1.5.1"
  }
}
```
We do have some info for the errored package, but we don't have its module or, more importantly, any of its dependencies.

So in this case, `rsc.io/sampler` not having a checksum causes us to miss the `golang.org/x/text/internal/tag` package (among others) entirely, because sampler was the only thing that depended on it.

> I'm assuming this won't happen because we merge the ouput of both `go list` and `go download` for module reporting, right?

Hmm, yeah, I think we will always report all the _modules_. We might be missing packages though.

**Comment by chmeliik** [comment]
Documented the problems with missing checksums (the additional problems, not just the reduced security)

**Comment by brunoapimentel** [comment]
> Hmm, yeah, I think we will always report all the _modules_. We might be missing packages though.

My main point of concern is if a repository with an incomplete `go.sum` would somehow get past the build stage and get shipped with an incomplete SBOM. If we are reporting all modules and we're pretty much sure the build would fail regardless, I think it's fine. There'll be the EC that might eventually block releases with missing checksums.

**Comment by lkolacek** [comment]
> There'll be the EC that might eventually block releases with missing checksums.
+1, I would expect EC will handle this.

Overall looks good Adam! I'll do 1 more detailed review tomorrow. :+1:

File: cachi2/core/package_managers/gomod.py [hunk 5]
```diff
@@ -448,9 +502,16 @@ def _find_missing_gomod_files(source_path: RootedPath, subpaths: list[str]) -> l
     return invalid_gomod_files
 
 
-def _resolve_gomod(
-    app_dir: RootedPath, request: Request, tmp_dir: Path
-) -> tuple[ParsedModule, Iterable[ParsedModule], Iterable[ParsedPackage]]:
+class ResolvedGoModule(NamedTuple):
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/gomod.py:505]: Nitpick: I'd still move this class to the top of the file alongside the other class definitions (although there is still one class we carried over from Cachito that is defined in the middle of functions :disappointed: ).

And the `Go` part of the name is probably redundant.

File: cachi2/core/package_managers/gomod.py [hunk 6]
```diff
@@ -448,9 +502,16 @@ def _find_missing_gomod_files(source_path: RootedPath, subpaths: list[str]) -> l
...
+    """Contains the data for a resolved Go module."""
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/gomod.py:506]: Maybe add a note to make it clear this is an input ~package~ module (specified by the user)? We already have several `*Module`s :grin:

**Review by brunoapimentel — APPROVED** [review]
Left two very minor comments. 

LGTM :rocket:

File: cachi2/core/package_managers/gomod.py (hunk 6 above)
Comment (brunoapimentel) [reply cachi2/core/package_managers/gomod.py:506]: Or maybe just call it ResolveResult, as it is being called in the variable names?

**Comment by brunoapimentel** [comment]
> My main point of concern is if a repository with an incomplete `go.sum` would somehow get past the build stage and get shipped with an incomplete SBOM. If we are reporting all modules and we're pretty much sure the build would fail regardless, I think it's fine. There'll be the EC that might eventually block releases with missing checksums.

We could consider adding a `.metatada.properties` item that tells the user an SBOM is/might be incomplete (similar to what OSBS does). Could be a good way to inform such cases, specially if we find more of them along the way.

File: cachi2/core/package_managers/gomod.py (hunk 6 above)
Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py]: I originally called it ResolveResult, but then updated it to match the [npm code ](https://github.com/containerbuildsystem/cachi2/blob/3f8752c89f46f3fbfe454a3fea564b7aa90f73d5/cachi2/core/package_managers/npm.py#L50)

Added the note though

**Comment by chmeliik** [comment]
> We could consider adding a `.metatada.properties` item that tells the user an SBOM is/might be incomplete (similar to what OSBS does). Could be a good way to inform such cases, specially if we find more of them along the way.

Something like a `cachi2:definitely_incomplete: true`, and when it's false or missing, assume that you don't know?

Whether cachi2 can assert that an SBOM is complete or not depends on how you want to use/interpret that SBOM. If you merge it with another SBOM like we do in RHTAP now, cachi2 can only tell you "definitely incomplete" or "IDK"

That probably still makes sense though. Want to file a design story?

**Comment by chmeliik** [comment]
And for Go modules, incomplete might as well be a constant `true`, because we list stdlib dependencies which are different for us vs. the user (what a feature that was)

**Comment by lkolacek** [comment]
> Whether cachi2 can assert that an SBOM is complete or not depends on how you want to use/interpret that SBOM. If you merge it with another SBOM like we do in RHTAP now, cachi2 can only tell you "definitely incomplete" or "IDK". (In fact it's a "probably", not definitely - missing checksums or other `go list` errors don't necessarily cause incompleteness)
> 
> That probably still makes sense though. Want to file a design story?

I would probably ask architects first for their opinion and then file a design story.

**Review by lkolacek — APPROVED** [review]
LGTM! :+1:

**Comment by brunoapimentel** [comment]
> And for Go modules, incomplete might as well be a constant `true`, because we list internal stdlib dependencies which are different for us vs. the user (what a feature that was)
> 
> (we would probably want to report what is incomplete and why)

Now I'm curious about that. I thought the stdlib packages varied only because of Go versions. Do they also change depending on the local environment?

**Comment by chmeliik** [comment]
> > And for Go modules, incomplete might as well be a constant `true`, because we list internal stdlib dependencies which are different for us vs. the user (what a feature that was)
> > (we would probably want to report what is incomplete and why)
> 
> Now I'm curious about that. I thought the stdlib packages varied only because of Go versions. Do they also change depending on the local environment?

Go versions, and maybe the platform. But keep in mind that the user can build with an entirely different Go version than the one we use to fetch and identify dependencies.

**Pull request merged**